### PR TITLE
Add friend tree capabilities to spark

### PR DESCRIPTION
**Pull request by vepadulano**
This is a barebones example of implementation to deal with
`TChain`s that have friend trees. The implementation follows the reasoning of
`GetFriendInfo` and `MakeChain` functions of the [TTreeProcessorMT](https://root.cern.ch/doc/master/classROOT_1_1TTreeProcessorMT.html) ROOT class.

The `RDataFrame` class is already capable of retrieveng the filenames of
the TTree/Tchain object passed as an argument to `PyRDF.RDataFrame()`.

Now it is also capable of retrieving treenames and filenames of friend trees,
if there are any. For the moment, the friend trees list is retrieved via
`TTree.GetListOfFriends()` and for each element in the list a tuple with
(treename, filename) is appended to a list.

During the map phase, in `Dist.py`, the list of tuples is retrieved and
added as attribute to each `Range` instance of the distributed execution,
as was previously done for the list of filenames of the `TTree`s in the
chain.

Finally, after reconstructing the `TChain`, the friend are added to the chain
via `TChain.AddFriend()`.

A test has been modified to also include the new attribute of the `Range` class

### CODE REVIEW

File: PyRDF/RDataFrame.py [hunk 1]
```diff
@@ -183,6 +183,29 @@ def get_inputfiles(self):
         # RDataFrame may have been created with no input files
         return None
 
+    def get_input_friend_treenames_and_filenames(self):
```
Comment (JavierCVilla) [inline PyRDF/RDataFrame.py]: Even though I can understand why you put it here, I would not implement this method in `RDataFrame.py` but in `Dist.py`. 

While the rest of `get` methods (`get_inputfiles`, `get_treename`, `get_branches` ... ) need to check the parameters passed to the `RDataFrame` constructor in order to access to the required information (input files, treename, branches,...), this method could be directly applied to a `TTree/TChain` as it is done in the ROOT implementation. Indeed, this method will be only called from a `Dist` class or subclass.

Comment (JavierCVilla) [inline PyRDF/RDataFrame.py]: Also I would use a shorter name, maybe the same as in the ROOT implementation: `get_friend_info`.

File: PyRDF/RDataFrame.py [hunk 2]
```diff
@@ -183,6 +183,29 @@ def get_inputfiles(self):
...
+        This list can be extracted if the argument passed to PyRDF.RDataFrame()
+        is a TChain.
+
+        Returns:
+            (list, None): list of tuples of (treename, filename) for each
+            friend, or None if there are no friend trees.
+
+        """
+        first_arg = self.args[0]
+        if isinstance(first_arg, ROOT.TChain) and first_arg.GetListOfFriends():
```
Comment (JavierCVilla) [inline PyRDF/RDataFrame.py]: Moving the method to `Dist` and getting a `TTree` as an argument, we can even remove these checks.

File: PyRDF/RDataFrame.py [hunk 3]
```diff
@@ -183,6 +183,29 @@ def get_inputfiles(self):
...
+            # Extract file names from a given TChain
+            chain = first_arg
+            return [
+                (friendElem.GetName(), friendElem.GetTitle())
```
Comment (JavierCVilla) [inline PyRDF/RDataFrame.py]: For the result we may want to be more explicit and use either namedtuples or a custom object to represent the `FriendInfo`.

File: PyRDF/backend/Dist.py [hunk 4]
```diff
@@ -209,7 +211,8 @@ def _get_clustered_ranges(self, nentries, npartitions, treename, filelist):
             offset_first_cluster = clusters[index_start][2]
             ranges.append(Range(start - offset_first_cluster,
                                 end - offset_first_cluster,
-                                range_files))
+                                range_files,
+                                friend_filelist))
```
Comment (JavierCVilla) [inline PyRDF/backend/Dist.py]: `range_friend_files`

File: PyRDF/backend/Dist.py [hunk 5]
```diff
@@ -306,6 +315,10 @@ def mapper(current_range):
                 for f in current_range.filelist:
                     chain.Add(f)
 
+                if current_range.friend_filelist:
+                    for f in current_range.friend_filelist:
+                        chain.AddFriend(f[0], f[1])
```
Comment (JavierCVilla) [inline PyRDF/backend/Dist.py]: This assumes that the friend trees only have one file, but it may also be a chain composed by many files.

First we need to recreate the same chain before adding it as friend, as in the root implementation.

Additionally, we cannot just add the friend, since we need to set the number of entries to be read, as we do for `chain` with `chain.SetCacheEntryRange`.

**Review by JavierCVilla — CHANGES_REQUESTED** [review]
Thank you @vepadulano for the quick PR after spotting the issue!

Proposed changes go on the right direction, however we need to rethink a bit the support for friend trees and reuse as much code as possible from the ROOT implementation as it solves the same issue.

File: tests/unit/backend/test_dist.py [hunk 6]
```diff
@@ -285,6 +285,7 @@ def test_buildranges_with_clustered_ranges(self):
         # Mock attributes accessed through self. inside build_ranges
         backend.treename = "myTree"
         backend.files = "tests/unit/backend/1000clusters.root"
+        backend.friend_filelist = None
```
Comment (JavierCVilla) [inline tests/unit/backend/test_dist.py]: Before I forget, we need a test with friend trees.

File: PyRDF/backend/Dist.py (hunk 4 above)
Comment (vepadulano) [reply PyRDF/backend/Dist.py]: I am in doubt about this one particularly. Thing is that `range_files` is created after retrieving the file name for that particular range of entries, but that doesn't automatically mean that for the same range of entries we would have the same friend file names, right? 
For the moment I am just passing all the filenames (and treenames actually) of the friend trees

Comment (JavierCVilla) [reply PyRDF/backend/Dist.py]: Right, this might require further investigation (also to see how it is currently implemented in root), as clusters in both sets of files may not be the same. 

`range_files` is added to the list of ranges for debugging purposes and to create `TChain`'s with a reduced size so we do not add unused files. At some point we would need to do the same for friend trees/chains but for now we can focus on the creation of the friend chain and add this optimization later on.

File: tests/integration/spark/test_include_headers.py [hunk 7]
```diff
@@ -8,11 +8,20 @@ class IncludesSparkTest(unittest.TestCase):
     Check that the required header files are properly included in Spark
     environment.
     """
+    def tear_down(self):
```
Comment (JavierCVilla) [inline tests/integration/spark/test_include_headers.py]: Can this be implemented with `tearDown` instead?

File: tests/integration/spark/test_include_headers.py [hunk 8]
```diff
@@ -8,11 +8,20 @@ class IncludesSparkTest(unittest.TestCase):
...
+        context = SparkContext.getOrCreate()
+        context.stop()
+
     def test_includes_function_with_filter_and_histo(self):
         """
         Check that the filter operation is able to use C++ functions that
         were included using header files.
         """
+        # Stop current SparkContext if present
+        self.tear_down()
```
Comment (JavierCVilla) [inline tests/integration/spark/test_include_headers.py]: This would not be needed if `tead_down` is implemented as `teadDown`

File: tests/integration/spark/test_inv_mass.py [hunk 9]
```diff
@@ -5,9 +5,7 @@
 
 class SparkHistogramsTest(unittest.TestCase):
     """Integration tests to check the working of PyRDF."""
-
-    @classmethod
-    def tearDown(cls):
+    def tear_down(self):
```
Comment (JavierCVilla) [inline tests/integration/spark/test_inv_mass.py]: What's the reason of this change?

**Review by JavierCVilla — APPROVED** [review]
